runtime/lua/vim/lsp/util.lua

```diff
@@ -396,6 +399,54 @@ function M.apply_text_edits(text_edits, bufnr, position_encoding)
     end
   end
 
+  --- Track how many times each change annotation is applied to build up the final description.
+  ---@type table<string, integer>
+  local change_count = {}
+
+  -- Apply text edits.
+  for _, text_edit in ipairs(text_edits) do
+    if text_edit.annotationId then
+      assert(
+        change_annotations ~= nil,
+        'change_annotations must be provided for annotated text edits'
+      )
+
+      local annotation = assert(
+        change_annotations[text_edit.annotationId],
+        string.format('No change annotation found for id: %s', text_edit.annotationId)
+      )
+
+      if annotation.needsConfirmation then
+        vim.ui.select(
+          { 'Yes', 'No' },
+          { prompt = string.format('Apply "%s"? ', annotation.label) },
+          function(input)
+            if input == 'Yes' then
+              change_count[text_edit.annotationId] = (change_count[text_edit.annotationId] or 0) + 1
+              apply_text_edit(text_edit)
+            end
+          end
+        )
```

Comment:
How does this behave if the `vim.ui.select` implementation is non-blocking and the text-edits loop continues?
Does it result in multiple popups and is that safe - in that one could apply them out of order?

Reply:
Hmm great point. The spec says that resource operations should be applied in the order in which they're provided, so if one of them requires confirmation and isn't applied we'll probably want to dismiss the remaining edits too.

Perhaps we should "batch" the confirmations into a single call to `vim.ui.select` with options for "Applying all" or "Dismiss", and then applying all the edits at once based on that. WDYT?

Reply:
Done :)